README.md

<h1 align="center">MOTS</h1>

<img align="center" src="img/logo.png"  alt="MOTS Logo">

<p align="center">
Simple tool for hosting static files and makes fakes APIs to test your apps.
Make with ❤️ and Typescript by Yoannis Sánchez Soto</p>

![npm](https://img.shields.io/npm/v/mots)
![downloads/month](https://img.shields.io/npm/dm/mots)
![Telegram](https://img.shields.io/badge/t.me/yossthedev-Telegram-BLUE?style=flat&logo=Telegram)
![Twitter](https://img.shields.io/twitter/follow/yossthedev?style=social)

<img src="img/screen.png"
align="center"
        alt="Picture"
        style="display: block; margin: 0 auto" />

## 🚀 Features

* Intuitive and easy to use commands
* CORS enabled by default
* Hosts multiple folders
* Makes fake APIs to test your apps **PREFABS**

## 📦 Prefabs

MOTS allows you to create fake apis called Prefabs, these are configurable files with the structure of an API able to respond to parameters and define a response based on this, it can be a file, an html page or anything.

This feature is intended to help developers to test their apps without the need to have developed an API before or without the need to be connected all the time.

### Creating Prefabs

To create a new Prefab you must enter the following command

    mots prefab create [NAME]
Just provide a name for the new Prefab.

It will indicate the folder on your computer where it has been saved. For example

    💾 Prefab Saved in /home/user/.local/share/mots/prefabs/new.json

This newly created prefab already has a default structure designed as a starting point, now try to serve it with the following command
  
     mots prefab [NAME]

Try entering in your browser the following url to see the change of the parameters **localhost:4004/?user=you**

All prefabs contain the following structure

``` json
{
  "port": 4004,
  "paths": [
    {
      "name": "",
      "folder": "",
      "response": "<h1>MOTS loves you $USER</h1><h2>Add ?user=[ANY] to url to change the bear for a name or anything else</h2>",
      "params": [{ "name": "user", "default": "🐼", "isRequired": false }],
      "status": 200
    },
    {
      "name": "/Error",
      "folder": "",
      "response": "<h1>Not Found</h1>",
      "status": 404
    },
    {
      "name": "/Success",
      "folder": "",
      "response": "<h1>Yes, you can 🔥</h1>",
      "status": 200
    }
  ]
}

```

As you can see they are json files, with different properties easy to understand for any programmer 😁

## 👥 Contribution

Any contribution is always welcome. Please let me know of any bugs you detect in MOTS.

## 👨‍💻 Usage
<!-- usage -->
```sh-session
$ npm install -g mots
$ mots COMMAND
running command...
$ mots (--version)
mots/0.0.8 linux-x64 node-v16.17.0
$ mots --help [COMMAND]
USAGE
  $ mots COMMAND
...
```
<!-- usagestop -->
### Commands
<!-- commands -->
* [`mots about`](#mots-about)
* [`mots config`](#mots-config)
* [`mots config datadir`](#mots-config-datadir)
* [`mots help [COMMAND]`](#mots-help-command)
* [`mots prefab PREFAB`](#mots-prefab-prefab)
* [`mots prefab create PREFAB`](#mots-prefab-create-prefab)
* [`mots serve [FOLDER]`](#mots-serve-folder)

## `mots about`

Show About

```
USAGE
  $ mots about

DESCRIPTION
  Show About

EXAMPLES
  $ mots about
```

_See code: [dist/commands/about/index.ts](https://github.com/yossTheDev/mots/blob/v0.0.8/dist/commands/about/index.ts)_

## `mots config`

Configuration of mots

```
USAGE
  $ mots config

DESCRIPTION
  Configuration of mots
```

_See code: [dist/commands/config/index.ts](https://github.com/yossTheDev/mots/blob/v0.0.8/dist/commands/config/index.ts)_

## `mots config datadir`

Show default hosted folder

```
USAGE
  $ mots config datadir

DESCRIPTION
  Show default hosted folder

EXAMPLES
  $ mots config datadir
  /home/USER/.local/share/mserv
```

## `mots help [COMMAND]`

Display help for mots.

```
USAGE
  $ mots help [COMMAND] [-n]

ARGUMENTS
  COMMAND  Command to show help for.

FLAGS
  -n, --nested-commands  Include all nested commands in the output.

DESCRIPTION
  Display help for mots.
```

_See code: [@oclif/plugin-help](https://github.com/oclif/plugin-help/blob/v5.1.12/src/commands/help.ts)_

## `mots prefab PREFAB`

Host Prefabs locally. Prefabs are Json files that contains the structure of an API

```
USAGE
  $ mots prefab [PREFAB]

ARGUMENTS
  PREFAB  Prefab to be hosted

DESCRIPTION
  Host Prefabs locally. Prefabs are Json files that contains the structure of an API

EXAMPLES
  $ mots prefab
```

_See code: [dist/commands/prefab/index.ts](https://github.com/yossTheDev/mots/blob/v0.0.8/dist/commands/prefab/index.ts)_

## `mots prefab create PREFAB`

Create a new prefab with the default structure in the Prefabs folder

```
USAGE
  $ mots prefab create [PREFAB]

ARGUMENTS
  PREFAB  Name of Prefab to be created

DESCRIPTION
  Create a new prefab with the default structure in the Prefabs folder

EXAMPLES
  $ mots prefab create
```

## `mots serve [FOLDER]`

Initialize Server

```
USAGE
  $ mots serve [FOLDER] [-p <value>]

ARGUMENTS
  FOLDER  (Optional) Folder to be hosted. If it is not provide default Public folder is served, write config datadir to
          show the default public directory location

FLAGS
  -p, --port=<value>  App PORT Ex:8080. If this is not defined default port has to be used (4000)

DESCRIPTION
  Initialize Server

EXAMPLES
  $ mots serve path/to/folder -p 8080
```

_See code: [dist/commands/serve/index.ts](https://github.com/yossTheDev/mots/blob/v0.0.8/dist/commands/serve/index.ts)_
<!-- commandsstop -->


CHANGELOG.md
# Changelog

## **V 0.0.5**

* First public Release

## **V 0.0.6**

* Now you can define **port** in the serve command

Ex:

``` bash
   mots serve path/to/folder -p 8080
```

* Now you can serve multiple folders using the serve command separating with ","

Ex:

``` bash
   mots serve path/to/folder1,path/to/folder2
```

## **V 0.0.7**

* Adds Prefabs System
* Improve console outputs
* Fix some bugs

## **V 0.0.8**

## General

* Some bug fixed
* Improve console outputs
* Clean code

### Prefabs

* Add Reload hability to serve prefabs
* Add $DATADIR property to Prefabs`s folder property. You can use this to define custom folders inside Mots data directory


src/services/prefabs.ts
// eslint-disable-next-line node/no-extraneous-import
import * as fs from 'fs-extra';
import { Express } from 'express';
import express = require('express');
import path = require('path');
import * as ip from 'ip';
import chalk from 'chalk';
import cors = require('cors');
import url from 'node:url';
import { Server } from 'node:http';

// #region
interface prefab {
	port: number;
	paths: endpoint[];
	cors: boolean;
}
interface endpoint {
	name: string;
	params?: param[];
	folder: string;
	status: number;
	type?: string;
	errorResponse?: string;
	response: string;
}
interface param {
	name: string;
	default: any;
	isRequired: boolean;
}
// #endregion

// eslint-disable-next-line valid-jsdoc
/**
 * Get Prefab Item
 * @param dataDir MOTS data dir
 * @param prefab  Prefab Name
 */
export async function getPrefab(
	dataDir: string,
	prefab: string,
): Promise<Server> {
	// Read Prefab File
	const prefabItem = (await fs.readJson(
		`${dataDir}/prefabs/${prefab}.json`,
	)) as prefab;

	// Create Express app
	const app: Express = express();
	const port = prefabItem.port;

	// Verify is CORS is used
	if (prefabItem.cors) {
		app.use(cors());
	}

	// Get all paths and serve Prefab item
	for (const path of prefabItem.paths) {
		// If folder property is provided, serve this folder
		if (path.folder) {
			// if folder property contains spacial varialbles
			// like $DATADIR, replace this
			if (path.folder.includes('$DATADIR')) {
				// Serve static folder
				app.use(express.static(path.folder.replace('$DATADIR', dataDir)));
			} else {
				// Serve static folder
				app.use(express.static(path.folder));
			}
		}

		// Create get endpoint
		app.get(`${path.name}/` || '/', (req, res) => {
			let response = path.response;

			if (path.params) {
				for (const param of path.params) {
					// Get parameter from url
					const resParam = url.parse(req.url, true).query[param.name];

					// Verify if this parameter is required
					if (param.isRequired) {
						if (resParam) {
							// If the parameter is reuired and te user submits it
							response = parseParamsInResponse(
								param.name,
								resParam as string,
								response,
							);
						} else {
							console.log();
							// If the parameter is required, but is not sent but has a default value
							// eslint-disable-next-line max-depth
							if (param.default) {
								response = parseParamsInResponse(
									param.name,
									param.default,
									response,
								);
							} else {
								// Parameter not send and not has a default value
								return res.status(path.status || 200).send(path.errorResponse);
							}
						}
					} else {
						// If the user does not send the parameter but has a default value,
						// this is taken as if it was the parameter sent by the user

						// Optional sended parameter
						response = resParam
							? parseParamsInResponse(param.name, resParam as string, response)
							: parseParamsInResponse(param.name, param.default, response);
					}
				}
			}

			if (path.type === 'redirect') {
				res.status(path.status || 200).redirect(response);
			} else {
				res.status(path.status || 200).send(response);
			}
		});
	}

	// App listen
	const server = app.listen(port);

	// Logging
	logPrefab(prefabItem.paths, port);

	return server;
}

// eslint-disable-next-line valid-jsdoc
/**
 *  Serve prefabs locally
 * @param dataDir Mots data dir
 * @param prefab Name of the prefab to be hosted
 *
 */
export async function servePrefab(
	dataDir: string,
	prefab: string,
): Promise<void> {
	// Get prefabs folder or create it
	fs.mkdirSync(`${dataDir}/prefabs`, { recursive: true });

	// Get all files in the folder
	const files = fs.readdirSync(path.join(dataDir, 'prefabs'));

	for (const item of files) {
		if (
			path.extname(item) === '.json' &&
			path.basename(item) === prefab + '.json'
		) {
			// Serve Prefab item
			// eslint-disable-next-line no-await-in-loop
			let app = await getPrefab(dataDir, prefab);

			fs.watchFile(`${dataDir}/prefabs/${prefab}.json`, async () => {
				// Notify Reloading
				console.log(`
				⚡️ ${chalk.bold('RELOADING SERVER')} ⚡️`);

				// Close previous server
				app.close();

				// Start new server
				app = await getPrefab(dataDir, prefab);
			});
		}
	}
}

// eslint-disable-next-line valid-jsdoc
/**
 * Create new prefab with default data
 * @param name Prefab Name
 * @param dataDir Mots Data Dir
 */
export async function createDefaultPrefab(
	name: string,
	dataDir: string,
): Promise<void> {
	// Get prefabs folder or create it
	fs.mkdirSync(`${dataDir}/prefabs`, { recursive: true });

	// Set default prefab data
	const defaultPrefab: prefab = {
		port: 4004,
		cors: true,
		paths: [
			{
				name: '',
				folder: '',
				response:
					'<h1>MOTS loves you $USER</h1><h2>Add ?user=[ANY] to url to change the bear for a name or anything else</h2>',
				params: [{ name: 'user', default: '🐼', isRequired: false }],
				status: 200,
			},
			{
				name: '/Error',
				folder: '',
				response: '<h1>Not Found</h1>',
				status: 404,
			},

			{
				name: '/Success',
				folder: '',
				response: '<h1>Yes, you can 🔥</h1>',
				status: 200,
			},
		],
	};

	// Save default prefab
	fs.writeJSON(
		path.join(dataDir, 'prefabs') + '/' + name + '.json',
		defaultPrefab,
	);
}

/**
 * Replace the value of the parameters in the response
 * Ex:
 * <h1>$ID</h1>
 * The $ID keyword will be replaced by the value that user has sent for the parameter
 * @param param Parameter name
 * @param value Value to replace
 * @param response Response
 * @returns Returns the answer with the parameters replaced
 */
function parseParamsInResponse(
	param: string,
	value: string,
	response: string,
): string {
	return response.replace(`$${param.toUpperCase()}`, value);
}

// eslint-disable-next-line valid-jsdoc
/**
 * Logging Prefab Item
 * @param paths Prefab Endpoints
 * @param port Port to be used
 */
async function logPrefab(paths: endpoint[], port: number) {
	// Log this
	console.log(`
	----------------------------			
	| ${chalk.bold(chalk.blue('⚡️ THE SERVER IS READY ⚡️'))}|
	----------------------------

  🌐 ${chalk.bold('HOSTS')} 

    -> Server is running at:
	-> http://localhost:${port}
	-> http://${ip.address()}:${port} 

  🚩 ${chalk.bold('ENDPOINTS')} ->	
	`);

	for (const i of paths) {
		console.log(`      -> http://localhost:${port}${i.name}`);
	}
}
